xapian-letor/ranker/listnet_ranker.cc

```diff
@@ -136,46 +132,56 @@ ListNETRanker::train_model() {
     // iterations
     for (int iter_num = 1; iter_num < iterations; ++iter_num) {
 	// initialize Probability distributions of y and z
-	prob_distrib_vector prob = initializeProbability(fvv, new_parameters);
+	prob_distrib_vector prob = initializeProbability(training_data, new_parameters);
 	// compute gradient
-	vector<double> gradient = calculateGradient(fvv, prob);
+	vector<double> gradient = calculateGradient(training_data, prob);
 	// update parameters: w = w - gradient * learningRate
 	updateParameters(new_parameters, gradient, learning_rate);
     }
     swap(parameters, new_parameters);
 }
 
 void
-ListNETRanker::save_model_to_file(const char* output_filename) {
-    LOGCALL_VOID(API, "ListNETRanker::save_model_to_file", output_filename);
-    ofstream parameters_file;
-    parameters_file.open(output_filename);
+ListNETRanker::save_model_to_metadata(const string & model_key) {
+    LOGCALL_VOID(API, "ListNETRanker::save_model_to_file", model_key);
+    Xapian::WritableDatabase letor_db(get_database_path());
+    string key = model_key;
+    if (key.empty()) {
+	key = "LisntNET.model.default";
+    }
+    ostringstream oss;
     for (size_t i = 0; i <  parameters.size(); ++i)
-	parameters_file << setprecision(numeric_limits<double>::digits10) << parameters[i] << endl;
-    parameters_file.close();
+	oss << setprecision(numeric_limits<double>::digits10) << parameters[i] << endl;
+    string model_data = oss.str();
+    letor_db.set_metadata(key, model_data);
```

Comment:
I'd suggest using `serialise_double()` here - see `common/serialise-double.h` (and `.cc`).

This would reliably store the values without any loss of precision (I'm not entirely convinced this code always will), but should result in a smaller string for a given model (`serialise_double()` will encode any value in at most 10 bytes, and many in less than that; an ASCII string of the decimal representation will be much longer on average, as it only encodes 10 possible values per byte).

I'd expect `serialise_double()` to be faster too, though that's not a big concern here.